### PR TITLE
feature/validation

### PR DESCRIPTION
**Pull request by ncocquio**
#8

### CODE REVIEW

File: src/components/ProjectModelForm/ProjectCreateForm.tsx [hunk 1]
```diff
@@ -51,14 +52,19 @@ export const ProjectCreateForm: Component<ProjectFormProps> = ({ state, actions
...
             />
+            {name.errors == null && <FormHint label={String.fromCharCode(160)} />}
+            {name.errors != null && name.errors.required && <FormHint theme="danger" label="Name is required" />}
           </FormField>
           <FormField labelText="FTE" required={true}>
             <FormInput
               name={ftePercentage.name}
               value={ftePercentage.value}
+              min={1}
+              max={9223372036854775807}
```
Comment (kelvinlouis) [inline src/components/ProjectModelForm/ProjectCreateForm.tsx]: Please use a constant

File: src/components/ProjectModelForm/ProjectCreateForm.tsx [hunk 2]
```diff
@@ -51,14 +52,19 @@ export const ProjectCreateForm: Component<ProjectFormProps> = ({ state, actions
             <FormInput
               name={name.name}
               value={name.value}
+              maxLength={50}
               type="text"
               onInputChange={formActions.updateValue}
             />
+            {name.errors == null && <FormHint label={String.fromCharCode(160)} />}
```
Comment (kelvinlouis) [inline src/components/ProjectModelForm/ProjectCreateForm.tsx]: Remove all placeholders and use min-height in css instead

File: src/components/FormHint/FormHint.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+export const FormHint: Component<Props> = ({ label, theme }) => {
+  let className = 'help';
+
+  if (theme) {
+    className = `${className} is-${theme}`;
+  }
+
+  return (
```
Comment (kelvinlouis) [inline src/components/FormHint/FormHint.tsx:13]: Please put everything on one line

File: src/components/EmployeeModalForm/EmployeeCreateForm.tsx [hunk 4]
```diff
@@ -49,25 +49,35 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
             <FormInput
               name={firstName.name}
               value={firstName.value}
+              maxLength={50}
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeCreateForm.tsx]: Pass errors to `FormInput`, `FormSelect`, `DatePicker`, `EmployeeSelect` and `FormCheckbox` and mark the control with `is-danger` if it contains errors.

Comment (kelvinlouis) [reply src/components/EmployeeModalForm/EmployeeCreateForm.tsx]: Add constants for 50 => `INPUT_MIN_LENGTH_SHORT = 50`

File: src/components/ContractForm/ContractForm.tsx [hunk 5]
```diff
@@ -45,19 +48,22 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
...
       <div className="contract-form__column">
-        <FormField labelText="Pensum" required={true} hint="0-100">
+        <FormField labelText="Pensum" required={true}>
           <FormInput
             name={pensumPercentage.name}
             value={pensumPercentage.value}
             type="number"
             suffix="fas fa-percent"
-            min={0}
+            min={1}
```
Comment (kelvinlouis) [inline src/components/ContractForm/ContractForm.tsx]: Add constant `PENSUM_VALUE_MIN` resp. `PENSUM_VALUE_MAX`.

File: src/components/EmployeeModalForm/EmployeeCreateForm.tsx [hunk 6]
```diff
@@ -49,25 +49,35 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
...
               onInputChange={formActions.updateValue}
             />
+            {firstName.errors == null && <FormHint label={String.fromCharCode(160)} />}
+            {firstName.errors != null && firstName.errors.required && <FormHint theme="danger" label="First name is required" />}
           </FormField>
           <FormField labelText="Last Name" required={true}>
             <FormInput
               name={lastName.name}
               value={lastName.value}
+              maxLength={50}
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeCreateForm.tsx]: Reuse the component from above.

File: src/components/EmployeeModalForm/EmployeeCreateForm.tsx [hunk 7]
```diff
@@ -49,25 +49,35 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
...
               onInputChange={formActions.updateValue}
             />
+            {lastName.errors == null && <FormHint label={String.fromCharCode(160)} />}
+            {lastName.errors != null && lastName.errors.required && <FormHint theme="danger" label="Last name is required" />}
           </FormField>
           <FormField labelText="Email" required={true}>
             <FormInput
               name={emailAddress.name}
               value={emailAddress.value}
+              maxLength={120}
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeCreateForm.tsx]: Add constants for 120 => `INPUT_MIN_LENGTH_MEDIUM = 120`

File: src/components/EmployeeModalForm/EmployeeCreateForm.tsx [hunk 8]
```diff
@@ -76,14 +86,17 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
...
           </FormField>
-          <FormField labelText="Status" required={true} hint="Inactive: No login possible">
+          <FormField labelText="Status" required={true}>
             <FormCheckbox
               name={active.name}
               value={active.value}
               labelText="Active"
               onInputChange={formActions.updateValue}
             />
+            {active.errors == null && <FormHint label={"Inactive: No login possible"} />}
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeCreateForm.tsx]: remove the braces

File: src/components/EmployeeModalForm/EmployeeEditForm.tsx [hunk 9]
```diff
@@ -50,44 +51,57 @@ export const EmployeeEditForm: Component<Props> = ({ state, actions }) => {
             <FormInput
               name={firstName.name}
               value={firstName.value}
+              maxLength={50}
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeEditForm.tsx]: Reuse constant

File: src/components/EmployeeModalForm/EmployeeEditForm.tsx [hunk 10]
```diff
@@ -50,44 +51,57 @@ export const EmployeeEditForm: Component<Props> = ({ state, actions }) => {
...
-              onInputChange={formActions.updateValue}
-            />
+            {lastName.errors == null && <FormHint label={String.fromCharCode(160)} />}
+            {lastName.errors != null && lastName.errors.required && <FormHint theme="danger" label="Field is required" />}
           </FormField>
           <FormField labelText="Email" required={true}>
             <FormInput
               name={emailAddress.name}
               value={emailAddress.value}
+              maxLength={120}
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeEditForm.tsx]: Reuse constant

File: src/components/ProjectModelForm/ProjectCreateForm.tsx [hunk 11]
```diff
@@ -51,14 +52,19 @@ export const ProjectCreateForm: Component<ProjectFormProps> = ({ state, actions
             <FormInput
               name={name.name}
               value={name.value}
+              maxLength={50}
```
Comment (kelvinlouis) [inline src/components/ProjectModelForm/ProjectCreateForm.tsx]: Reuse constant

File: src/components/ProjectModelForm/ProjectCreateForm.tsx [hunk 12]
```diff
@@ -51,14 +52,19 @@ export const ProjectCreateForm: Component<ProjectFormProps> = ({ state, actions
...
               onInputChange={formActions.updateValue}
             />
+            {name.errors == null && <FormHint label={String.fromCharCode(160)} />}
+            {name.errors != null && name.errors.required && <FormHint theme="danger" label="Name is required" />}
           </FormField>
           <FormField labelText="FTE" required={true}>
             <FormInput
               name={ftePercentage.name}
               value={ftePercentage.value}
+              min={1}
```
Comment (kelvinlouis) [inline src/components/ProjectModelForm/ProjectCreateForm.tsx]: Reuse constant

File: src/state/form/employee-form.state.ts [hunk 13]
```diff
@@ -23,22 +24,38 @@ export const initEmployeeForm: () => EmployeeFormState = () => ({
...
       value: null,
+      validators: [
+        textRequiredValidator,
+      ],
     },
     role: {
       name: 'role',
       value: null,
+      validators: [
+        selectValidator("Please select"),
```
Comment (kelvinlouis) [inline src/state/form/employee-form.state.ts]: Single quote

File: src/state/form/validators.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+      negativeDuration: true,
+    };
+  }
+
+  return;
+};
+
+export const selectValidator = (defaultValue: string) => (control: FormControl<string>, state: BaseForm): FormErrors => {
+
+  if (control.value != null && control.value === defaultValue) {
```
Comment (kelvinlouis) [inline src/state/form/validators.ts]: Maybe use `value` for the placeholder option. This way you could ignore `defaultValue` as a parameter.

**Review by kelvinlouis — CHANGES_REQUESTED** [review]

File: src/components/ContractForm/ContractForm.tsx [hunk 15]
```diff
@@ -35,6 +37,8 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
             max={endDate.value}
             onInputChange={updateContractFormValue(key, formActions)}
           />
+          {startDate.errors == null && <FormHint label={String.fromCharCode(160)} />}
```
Comment (kelvinlouis) [inline src/components/ContractForm/ContractForm.tsx]: You shall remove

File: src/components/ContractForm/ContractForm.tsx [hunk 16]
```diff
@@ -45,19 +49,22 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
             min={startDate.value}
             onInputChange={updateContractFormValue(key, formActions)}
           />
+          {endDate.errors == null && <FormHint label={String.fromCharCode(160)} />}
```
Comment (kelvinlouis) [inline src/components/ContractForm/ContractForm.tsx]: You shall remove

File: src/components/ContractForm/ContractForm.tsx [hunk 17]
```diff
@@ -45,19 +49,22 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
...
             value={pensumPercentage.value}
             type="number"
             suffix="fas fa-percent"
-            min={0}
-            max={100}
+            min={CONTRACT_PENSUM_VALUE_MIN}
+            max={CONTRACT_PENSUM_VALUE_MAX}
             onInputChange={updateContractFormValue(key, formActions)}
           />
+          {pensumPercentage.errors == null && <FormHint label={String.fromCharCode(160)} />}
```
Comment (kelvinlouis) [inline src/components/ContractForm/ContractForm.tsx]: You shall remove

File: src/components/ContractForm/ContractForm.tsx [hunk 18]
```diff
@@ -45,19 +49,22 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
...
-        <FormField labelText="Pensum" required={true} hint="0-100">
+        <FormField labelText="Pensum" required={true}>
           <FormInput
             name={pensumPercentage.name}
             value={pensumPercentage.value}
             type="number"
             suffix="fas fa-percent"
-            min={0}
-            max={100}
+            min={CONTRACT_PENSUM_VALUE_MIN}
```
Comment (kelvinlouis) [inline src/components/ContractForm/ContractForm.tsx:65]: you might want to add errors here too.

File: src/components/EmployeeModalForm/EmployeeCreateForm.tsx [hunk 19]
```diff
@@ -49,41 +50,54 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
...
             />
+            {lastName.errors != null && lastName.errors.required && <FormHint theme="danger" label="Last name is required" />}
           </FormField>
           <FormField labelText="Email" required={true}>
             <FormInput
               name={emailAddress.name}
               value={emailAddress.value}
+              maxLength={INPUT_LENGTH_LONG_MAX}
               type="email"
+              hasError={emailAddress.errors != null && (emailAddress.errors.required || emailAddress.errors.email)}
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeCreateForm.tsx]: emailAddress.errors != null should suffice.

File: src/components/EmployeeModalForm/EmployeeCreateForm.tsx [hunk 20]
```diff
@@ -49,41 +50,54 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
...
             />
+            {emailAddress.errors != null && emailAddress.errors.required && <FormHint theme="danger" label="Email is required" />}
+            {emailAddress.errors != null && emailAddress.errors.email && <FormHint theme="danger" label="Email invalid" />}
           </FormField>
           <FormField labelText="Password" required={true}>
             <FormInput
               name={password.name}
               value={password.value}
               type="password"
+              hasError={password.errors != null && password.errors.required}
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeCreateForm.tsx]: password.errors != null should suffice

File: src/components/EmployeeModalForm/EmployeeCreateForm.tsx [hunk 21]
```diff
@@ -49,41 +50,54 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
...
             />
+            {firstName.errors != null && firstName.errors.required && <FormHint theme="danger" label="First name is required" />}
           </FormField>
           <FormField labelText="Last Name" required={true}>
             <FormInput
               name={lastName.name}
               value={lastName.value}
+              maxLength={INPUT_LENGTH_SHORT_MAX}
               type="text"
+              hasError={lastName.errors != null && lastName.errors.required}
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeCreateForm.tsx]: i would change `hasError` to `errors` and pass it the entire `errors` object. The `errors` object will be undefined, if there are no errors.

File: src/components/ContractForm/ContractForm.tsx (hunk 15 above)
Comment (kelvinlouis) [reply src/components/ContractForm/ContractForm.tsx]: Make sure you pass the errors object to the FormControl

File: src/components/ContractForm/ContractForm.tsx (hunk 16 above)
Comment (kelvinlouis) [reply src/components/ContractForm/ContractForm.tsx]: Add the errors to the form control

File: src/components/EmployeeModalForm/EmployeeCreateForm.tsx [hunk 22]
```diff
@@ -49,41 +50,54 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
...
               name={password.name}
               value={password.value}
               type="password"
+              hasError={password.errors != null && password.errors.required}
               onInputChange={formActions.updateValue}
             />
+            {password.errors != null && password.errors.required && <FormHint theme="danger" label="Password is required" />}
           </FormField>
-          <FormField labelText="Status" required={true} hint="Inactive: No login possible">
+          <FormField labelText="Status" required={true}>
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeCreateForm.tsx:97]: You might want to add the errors to the FormCheckbox too.

File: src/components/EmployeeModalForm/EmployeeEditForm.tsx [hunk 23]
```diff
@@ -50,44 +52,57 @@ export const EmployeeEditForm: Component<Props> = ({ state, actions }) => {
...
+              maxLength={INPUT_LENGTH_LONG_MAX}
               type="email"
+              hasError={emailAddress.errors != null && (emailAddress.errors.required || emailAddress.errors.email)}
               onInputChange={formActions.updateValue}
             />
+            {emailAddress.errors != null && emailAddress.errors.required && <FormHint theme="danger" label="Field is required" />}
+            {emailAddress.errors != null && emailAddress.errors.email && <FormHint theme="danger" label="Email invalid" />}
           </FormField>
-          <FormField labelText="Status" required={true} hint="Inactive: No login possible">
+          <FormField labelText="Status" required={true}>
```
Comment (kelvinlouis) [inline src/components/EmployeeModalForm/EmployeeEditForm.tsx:88]: Add errors here too.

File: src/components/FormControlProps.ts [hunk 24]
```diff
@@ -5,4 +5,5 @@ export interface FormControlProps<T> extends FormControl<T> {
   placeholder?: string;
   disabled?: boolean;
   isLoading?: boolean;
+  hasError?: boolean;
```
Comment (kelvinlouis) [inline src/components/FormControlProps.ts]: pass it the FormErrors type:

`errors: FormErrors`

File: src/components/FormInput/FormInput.tsx [hunk 25]
```diff
@@ -20,7 +22,9 @@ const FormInputSuffix: Component<FormInputSuffixProps> = ({ suffix }) => (
 
 export const FormInput: Component<Props> = (props) => {
   const controlClassName = props.suffix ? 'control has-icons-right' : 'control';
-  const inputClassName = props.isLoading ? 'input is-loading' : 'input';
+  let inputClassName = 'input';
+  inputClassName = props.isLoading ? inputClassName + ' is-loading' : inputClassName;
+  inputClassName = props.hasError ? inputClassName + ' is-danger' : inputClassName;
```
Comment (kelvinlouis) [inline src/components/FormInput/FormInput.tsx]:
```
inputClassName = props.errors != null ? inputClassName + ' is-danger' : inputClassName;
```

File: src/components/FormSelect/FormSelect.tsx [hunk 26]
```diff
@@ -52,6 +52,8 @@ export const FormSelect: Component<FormSelectProps<any, any>> = (props) => {
     value = o => o != null ? props.valueMapper!(o) : null;
   }
 
+  const selectClassName = props.hasError ? 'select is-danger' : 'select';
```
Comment (kelvinlouis) [inline src/components/FormSelect/FormSelect.tsx]: const selectClassName = props.errors != null ? 'select is-danger' : 'select';

**Review by kelvinlouis — CHANGES_REQUESTED** [review]

File: src/components/EmployeeModalForm/EmployeeCreateForm.tsx (hunk 22 above)
Comment (ncocquio) [reply src/components/EmployeeModalForm/EmployeeCreateForm.tsx:97]: Checkbox has no styling options

File: src/components/ContractForm/ContractForm.tsx (hunk 18 above)
Comment (ncocquio) [reply src/components/ContractForm/ContractForm.tsx:65]: We eliminated the textLengthValidator in favor of HTML input limits.
Spinner has limits set, so a validator is obsolete.

**Pull request merged**